src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -3781,17 +3781,16 @@ internal virtual BoundExpressionStatement BindConstructorInitializer(Constructor
                 // The constructor is implicit. We need to get the binder for the body
                 // of the enclosing class.
                 CSharpSyntaxNode containerNode = constructor.GetNonNullSyntaxNode();
-                BinderFactory binderFactory = compilation.GetBinderFactory(containerNode.SyntaxTree);
 
-                if (containerNode is TypeDeclarationSyntax typeDecl)
+                if (containerNode is CompilationUnitSyntax)
                 {
-                    outerBinder = binderFactory.GetInTypeBodyBinder(typeDecl);
-                }
-                else
-                {
-                    SyntaxToken bodyToken = GetImplicitConstructorBodyToken(containerNode);
-                    outerBinder = binderFactory.GetBinder(containerNode, bodyToken.Position);
+                    // Must be a source of top level statements with a partial type declaration
+                    // that specifies non-object base. The object base is handled above.
+                    containerNode = containingType.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).OfType<TypeDeclarationSyntax>().First();
```

Comment:
This comment implies that we need to be filtering for the type declaration that actually has the base type declaration, which is not what this code actually does. Is it worth adding another test for a `Program` with 3 parts: the top-level statement, an empty class body, and a body that declares the base type?

Reply:
> The comment explains under what conditions we can get here. It doesn't say what to do, or what filtering to apply.

That isn't how I'm reading it though. I find it confusing and I think we need to change the wording to be clearer. My suggestion would be to add a sentence that states something like `We need an actual TypeDeclarationSyntax in order to locate the correct binder for this case.`